File: frontend/src/types/reports.ts
/**
 * Analysis report TypeScript types.
 *
 * Mirror the backend schemas in `backend/app/schemas/reports.py` for the
 * Agent-driven report module: generating a structured Analysis_Report for an
 * analysis session and downloading the exported report file (PDF / Word).
 * Supports inline report generation and download from the analysis dashboard
 * (Req 5.5-5.7).
 */

/** Exported report format. Only PDF and Word (.docx) are supported (Req 5.5). */
export type ReportFormat = 'pdf' | 'docx';

/** Human-readable label for a report format (Req 5.5). */
export const REPORT_FORMAT_LABELS: Record<ReportFormat, string> = {
  pdf: 'PDF',
  docx: 'Word',
};

/**
 * Request body for `POST /reports/generate` (Req 5.1-5.7).
 *
 * `sections` optionally restricts/orders the report sections (the five required
 * sections are always present regardless). `format` is an optional hint for the
 * export format(s) to produce.
 */
export interface GenerateReportRequest {
  analysis_id: string;
  sections?: string[];
  format?: ReportFormat;
}

/**
 * Response from `POST /reports/generate` (Req 5.1-5.4, 5.7).
 *
 * Exposes the structured report `content` alongside boolean download
 * availability flags so the client knows which formats can be downloaded via
 * the download endpoint (Req 5.5).
 */
export interface ReportResponse {
  id: string;
  session_id: string;
  title: string;
  content: Record<string, unknown>;
  has_pdf: boolean;
  has_docx: boolean;
  created_at: string;
}

/** Response from `GET /reports/{report_id}/download` (Req 5.5). */
export interface ReportDownloadResponse {
  download_url: string;
  format: string;
}

File: backend/app/schemas/reports.py
"""Report request/response schemas.

Pydantic models for the Agent-driven report-generation API (generate a report
for an analysis, and download an exported report file). Mirrors the
response-model style in ``app/schemas/analysis.py``
(``model_config = {"from_attributes": True}``).

The structured report ``content`` (data summary, key findings, statistical
results, visualizations, recommendations + data-source metadata) is produced by
the self-developed report-generation MCP via the Agent and persisted verbatim;
the API additionally tracks which export formats (PDF / Word) are available for
download.

Requirements: 5.1-5.7
"""

import uuid
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


class GenerateReportRequest(BaseModel):
    """Request body for generating a report from an analysis session.

    ``sections`` optionally restricts/orders the report sections (the five
    required sections are always present regardless). ``format`` is an optional
    hint forwarded to the Agent for the export format(s) to produce.
    """

    analysis_id: uuid.UUID
    sections: Optional[list[str]] = None
    format: Optional[str] = None


class ReportResponse(BaseModel):
    """Response schema for a generated report.

    Exposes the structured report ``content`` alongside boolean download
    availability flags (rather than raw S3 keys) so clients know which formats
    can be downloaded via the download endpoint (Requirement 5.5).
    """

    id: uuid.UUID
    session_id: uuid.UUID
    title: str
    content: dict[str, Any]
    has_pdf: bool
    has_docx: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class ReportDownloadResponse(BaseModel):
    """Response schema for a report download request.

    Returns a presigned URL the client can use to download the exported report
    file from S3, together with the resolved format.
    """

    download_url: str
    format: str
